### ISSUE TITLE
Minimize manual type annotations

### ISSUE THREAD

**Issue by joshsh** [CategoricalData/hydra#119]
Hydra's kernel sources are currently littered with type annotations, due to previous limitations of Hydra type inference. We can't remove all of the annotations (after #103 is landed) because there are still cases (polymorphic recursion) where annotations are needed. However, most of them can be removed. Do this after #103 has been merged into main.

**Comment by joshsh** [comment]
All done. The only manual type annotations remaining are for kludging `Eq =>` and `Ord =>` in generated Haskell. Removing these is the subject of #164.
